temporian/implementation/numpy/operators/combine.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2021 Google LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Implementation for the Combine operator."""
+
+
+from typing import Dict, List, Set
+import numpy as np
+
+from temporian.implementation.numpy.data.event_set import (
+    IndexData,
+    EventSet,
+    IndexType,
+)
+from temporian.core.operators.combine import Combine
+from temporian.core.operators.combine import (
+    FROM_FIRST,
+    FROM_INTERSECT,
+    FROM_UNION,
+)
+from temporian.implementation.numpy import implementation_lib
+from temporian.implementation.numpy.operators.base import OperatorImplementation
+
+
+class CombineNumpyImplementation(OperatorImplementation):
+    def __init__(self, operator: Combine) -> None:
+        assert isinstance(operator, Combine)
+        super().__init__(operator)
+
+    def __call__(self, **inputs: EventSet) -> Dict[str, EventSet]:
+        assert isinstance(self.operator, Combine)
+        output_schema = self.output_schema("output")
+        output_evset = EventSet(data={}, schema=output_schema)
+        index_from = self.operator.index_from
+
+        input_evsets = list(inputs.values())
+        first_input = input_evsets[0]
+        first_features = first_input.schema.feature_names()
+
+        # What index values to use
+        if index_from == FROM_FIRST:
+            combined_keys = first_input.data.keys()
+        elif index_from == FROM_UNION:
+            combined_keys = set()
+            for evset in input_evsets:
+                combined_keys |= set(evset.data.keys())
+        elif index_from == FROM_INTERSECT:
+            combined_keys = set(first_input.data.keys())
+            for evset in input_evsets[1:]:
+                combined_keys &= set(evset.data.keys())
+        else:
+            raise ValueError(f"Unknown parameter for combine: {index_from=}")
+
+        # Concatenate timestamps and features for each index, and sort
+        for index_key in combined_keys:
+            all_timestamps: List[np.ndarray] = []
+            all_feats: Dict[str, List[np.ndarray]] = {
```

Comment:
Should we index features by feature index?

Reply:
Yep, you're right. I'll use a list instead of dict with first_features' indexes.